components/TextInput.js

```diff
@@ -1,35 +1,88 @@
-import PropTypes from 'prop-types';
 import React, { PureComponent } from 'react';
+import autoBindReact from 'auto-bind/react';
+import PropTypes from 'prop-types';
 import { TextInput as RNTextInput } from 'react-native';
 
-import { connectStyle } from '@shoutem/theme';
 import { connectAnimation } from '@shoutem/animation';
+import { connectStyle } from '@shoutem/theme';
+
+import { Caption } from './Text';
+import { View } from './View';
 
 class TextInput extends PureComponent {
-  render() {
-    const { props } = this;
-    const style = {
-      ...props.style,
+  constructor(props) {
+    super(props);
+
+    autoBindReact(this);
+
+    this.state = {
+      isFocused: props.highlightWhenFocused && props.autoFocus,
     };
-    delete style.placeholderTextColor;
-    delete style.selectionColor;
-    delete style.underlineColorAndroid;
+  }
+
+  handleFocus() {
+    this.setState({ isFocused: true });
+  }
+
+  handleBlur() {
+    this.setState({ isFocused: false });
+  }
+
+  resolveProps() {
+    const { highlightWhenFocused, errorMessage, style } = this.props;
+    const { isFocused } = this.state;
+
+    let resolvedStyle = {
+      ...style,
+      borderWidth: (isFocused || !!errorMessage) ? 1 : 0,
+    }
+
+    if (!!errorMessage) {
+      resolvedStyle.borderColor = style.errorBorderColor;
+    }
+
+    delete resolvedStyle.placeholderTextColor;
+    delete resolvedStyle.selectionColor;
+    delete resolvedStyle.underlineColorAndroid;
+
+    if (highlightWhenFocused) {
+      return {
+        onFocus: () => this.handleFocus(),
```

Comment:
no need for arrow fns here. Just use shorthand syntax -> onFocus: this.handleFocus,

Reply:
Come to think of it, do we really need to conditionally pass the focus and blur handlers here ? Have the handlers do this logic, and just always pass the handlers, which will decide to react or not.

generally we don't like the dynamic prop calculation that happens on every render. Sometimes we use it for style, but i would limit this concept to style calculation only. Even then, i would prefer style resolution inside render, like this

style={[style.baseStyle, !!errorMessage && style.errorBorder, isFocused && style.focusedBorder]} ... It's easier to read, and it would use existent styles conditionally. Provided that they are created with StyleSheet, it would bring an additional benefit of optimization, whereas we're creating a new style object on every render here.

Reply:
I sincerely disagree that it's easier to read, it also doesn't take into account `highlightOnFocus` prop.

If I'm understanding your comment about optimization correctly, it doesn't apply to our use case or how we work.

Reply:
it was just an example. You should fortify the condition based on your case. 

The fact that we're not passing our styles through StyleSheet doesn't mean we won't in the future. The fact that we're avoiding doing things properly because of the provided reason only makes it harder to implement the improvement in the future, as we keep piling on the code that will be harder to change.